### ISSUE TITLE
Filecache tests failing intermittently

### ISSUE THREAD

**Issue by takluyver** [European-XFEL/EXtra-data#45]
cc @egorsobolev 

The new tests from PR #25 failed in [one recent CI run](https://travis-ci.org/github/European-XFEL/EXtra-data/jobs/670164016). It was only the Python 3.7 job that failed, but I suspect this is a random failure rather than something related to the Python version.

Here are the failure details:

```
_____________________________ test_filecache_large _____________________________

mock_spb_raw_run = '/tmp/tmp3z89k4l6'
filecache_512 = <extra_data.filecache.FileCache object at 0x7f27749dcc50>

    def test_filecache_large(mock_spb_raw_run, filecache_512):
        fc = filecache_512
    
        files = [os.path.join(mock_spb_raw_run, f) \
                 for f in os.listdir(mock_spb_raw_run) if f.endswith('.h5')]
        run = DataCollection.from_paths(files)
    
        trains_iter = run.trains()
        tid, data = next(trains_iter)
        assert tid == 10000
        device = 'SPB_IRU_CAM/CAM/SIDEMIC:daqOutput'
        assert device in data
        assert data[device]['data.image.pixels'].shape == (1024, 768)
>       assert len(fc._cache) == len(run.files)
E       assert 17 == 18
E         -17
E         +18

extra_data/tests/test_filecache.py:42: AssertionError
_____________________________ test_filecache_small _____________________________

mock_spb_raw_run = '/tmp/tmp3z89k4l6'
filecache_3 = <extra_data.filecache.FileCache object at 0x7f27749dcc50>

    def test_filecache_small(mock_spb_raw_run, filecache_3):
        fc = filecache_3
    
        files = [os.path.join(mock_spb_raw_run, f) \
                 for f in os.listdir(mock_spb_raw_run) if f.endswith('.h5')]
        run = DataCollection.from_paths(files)
        trains_iter = run.trains()
    
        for i in range(3):
>           tid, data = next(trains_iter)

extra_data/tests/test_filecache.py:56: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
extra_data/reader.py:1406: in __iter__
    yield tid, self._assemble_data(tid)
extra_data/reader.py:1397: in _assemble_data
    source_data[key] = ds[first : first + count]
h5py/_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py/_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/h5py/_hl/dataset.py:566: in __getitem__
    if len(mshape) < len(self.shape):
h5py/_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py/_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/h5py/_hl/dataset.py:286: in shape
    return self.id.shape
h5py/h5d.pyx:132: in h5py.h5d.DatasetID.shape.__get__
    ???
h5py/h5d.pyx:133: in h5py.h5d.DatasetID.shape.__get__
    ???
h5py/_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py/_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
>   ???
E   ValueError: Not a dataset (not a dataset)

h5py/h5d.pyx:289: ValueError
```

**Comment by egorsobolev** [comment]
Yes, unclear. Does a random failure mean a non-reproducible one? I will look at this.

**Comment by takluyver** [comment]
Thanks! Hopefully we can reproduce it at least some of the time.

I'd guess this is somehow due garbage collection running and closing the file between when we check the cached dataset and when we try to use it. Maybe we should follow h5py and just let the file object go out of scope (closing it when no objects from it remain) rather than explicitly closing it (invalidating all objects from it).

**Comment by egorsobolev** [comment]
The first test fails because the file list is forming in random order by `DataCollection.from_paths()` constructor in multiprocessing mode.

The second test fails because many FileAccess instances remain undestroyed after the failure of the first test. For the second test global file cache replace with the new instance and the test open creates new FileAccess instances for the same files. Then garbage collector destroys FileAccess instances remained from the first test. And they close the files with the same names in a new cache.

**Comment by egorsobolev** [comment]
The second issue resolves by linking `FileAccess` instances to the global `FileCache` instance actual on creation time

**Comment by egorsobolev** [comment]
The first issue does not affect functionality itself but only on the test. I can estimate the proper number of files stored in the cache or remove this check. But I would suggest sorting files. It is maybe better for sequential reading.

**Comment by takluyver** [comment]
Addressed by #48.
